#### Binaries

I needed to add `profobuf` to `nativeBuildInputs` (for `protoc` which is needed to compile the `prost-build` crate). After that the binary could be built, but the tests not yet.

#### Tests

I added `llvm` and `llvmPackages.libclang.lib` and `rocksdb` to `checkInputs`, and the following `preCheck`:
```nix
  preBuild = ''
    export LLVM_CONFIG_PATH="${llvm}/bin/llvm-config";
    export LIBCLANG_PATH="${llvmPackages.libclang.lib}/lib";
    # Used by build.rs in the rocksdb-sys crate. If we don't set these, it would
    # try to build RocksDB from source.
    export ROCKSDB_INCLUDE_DIR="${rocksdb}/include";
    export ROCKSDB_LIB_DIR="${rocksdb}/lib";
  '';
```

This made some progress, but now building `rocksdb-sys` fails:

```
error: failed to run custom build command for `librocksdb-sys v6.20.3`

Caused by:
  process didn't exit successfully: `/build/source/target/release/build/librocksdb-sys-966b98292a2d40a1/build-script-build` (exit status: 101)
  --- stderr
  /nix/store/9bps2z67qs1mghyvk8pb6qxpwf8jpfby-rocksdb-6.26.1/include/rocksdb/c.h:65:10: fatal error: 'stdarg.h' file not found
  /nix/store/9bps2z67qs1mghyvk8pb6qxpwf8jpfby-rocksdb-6.26.1/include/rocksdb/c.h:65:10: fatal error: 'stdarg.h' file not found, err: true
  thread 'main' panicked at 'unable to generate rocksdb bindings: ()', /build/solana-testnet-cli-1.9.2-vendor.tar.gz/librocksdb-sys/build.rs:44:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

That header is provided in `${llvmPackages.libclang.lib}/lib/clang/${llvmPackages.libclan.version}/include`, but adding that to the include path seems like the wrong way to go, I suspect it’s not using the right compiler, but I don’t really have time to go down this rabbit hole right now. Maybe for now it’s best to just disable the tests.

#### Package name

> let me know what you think about having solana-mainnet and solana-testnet rather than the versions.

I have no strong opinion on this, I we have both numbered and named cases in Nixpkgs already (e.g. `python{38,39}`, but also `nginx{Stable,Mainline}`. In all cases we do also have one without the suffix, so I would also add a `solana` package that is an alias for `solana-mainnet`.

--- a/pkgs/applications/blockchains/solana/default.nix
+++ b/pkgs/applications/blockchains/solana/default.nix
@@ -1,0 +1,56 @@
+{ stdenv
+, fetchFromGitHub
+, lib
+, rustPlatform
+, IOKit
+, Security
+, AppKit
+, pkg-config
+, udev
+, zlib
+, protobuf
+}:
+rustPlatform.buildRustPackage rec {
+  pname = "solana-testnet-cli";
+  version = "1.9.2";
+
+  src = fetchFromGitHub {
+    owner = "solana-labs";
+    repo = "solana";
+    rev = "v${version}";
+    sha256 = "sha256-wrv35vBohLztMZPb6gfZdCaXcjj/Y7vnQqINaI6dBM4=";
+  };
+
+  cargoSha256 = "sha256-A5uVa+cRmrkVyw7MFH4QAr0VIFi18wcc2VPFvQyT9EM=";
+
+  buildAndTestSubdir = "cli";
+
+  nativeBuildInputs = lib.optionals stdenv.isLinux [ protobuf pkg-config ];
+  buildInputs = lib.optionals stdenv.isLinux [ udev zlib ] ++ lib.optionals stdenv.isDarwin [ IOKit Security AppKit ];
+
+  # check phase fails
+  # on darwin with missing framework System. This framework is not available in nixpkgs
+  # on linux with some librocksdb-sys compilation error
+  doCheck = false;
+
+  # all the following are needed for the checkphase
+  # checkInputs = lib.optionals stdenv.isDarwin [ pkg-config rustfmt ];
+  # Needed to get openssl-sys to use pkg-config.
+  # OPENSSL_NO_VENDOR = 1;
+  # OPENSSL_LIB_DIR = "${openssl.out}/lib";
+  # OPENSSL_DIR="${lib.getDev openssl}";
+  # LLVM_CONFIG_PATH="${llvm}/bin/llvm-config";
+  # LIBCLANG_PATH="${llvmPackages.libclang.lib}/lib";
+  # Used by build.rs in the rocksdb-sys crate. If we don't set these, it would
+  # try to build RocksDB from source.
+  # ROCKSDB_INCLUDE_DIR="${rocksdb}/include";
+  # ROCKSDB_LIB_DIR="${rocksdb}/lib";
+
+  meta = with lib; {
+    description = "Web-Scale Blockchain for fast, secure, scalable, decentralized apps and marketplaces. ";
+    homepage = "https://solana.com";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ happysalada ];
+    platforms = platforms.unix;
+  };
+}

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -30042,6 +30042,10 @@ with pkgs;
 
   sumokoin = callPackage ../applications/blockchains/sumokoin { boost = boost165; };
 
+  solana-testnet = callPackage ../applications/blockchains/solana {
+    inherit (darwin.apple_sdk.frameworks) IOKit Security AppKit;
+  };
+
   tessera = callPackage ../applications/blockchains/tessera { };
 
   vertcoin  = libsForQt514.callPackage ../applications/blockchains/vertcoin {

